### ISSUE TITLE
Window width affects directional shadow resolution.

### ISSUE THREAD

**Issue by EzraT** [godotengine/godot#23603]
**Godot version:**
Affects both 3.0.6 and Official 3.1 Alpha 2

**OS/device including version:**
Linux Mint 19 - 64bit
Nvidia GeForce GTX 1060 on driver version: 396.54

**Issue description:**
For some reason the directional shadow resolution decreases as the window width increases, and increases as the width decreases.
Happens in the editor window and also in a running game.
[Video](https://github.com/godotengine/godot/files/2563615/what.flv.zip)

I'm assuming this isn't intended behaviour?
OmniLight and SpotLight do not seem affected.

**Steps to reproduce:**
1. Set up a scene with meshes and a Directional Light with shadows enabled.
2. Decrease and increase the width of the 3D scene window and observe shadows.

**Comment by EzraT** [comment]
Any chance this one can be added to the 3.1 milestone?
The quality difference is quite dramatic.

**Comment by and3rson** [comment]
Bumping as this issue still persists.

**Comment by Calinou** [comment]
I'm not sure if this can be resolved completely (any ideas?). In any case, we could make it so the shadow quality becomes worse when the viewport becomes taller, instead of wider. This would play better with ultrawide displays.

**Comment by Calinou** [comment]
I tested how shadow quality changes when the camera is set to use "Keep Width" (Vert-) aspect instead of "Keep Height" (Hor+). The shadow quality still decreases as the viewport becomes wider (even though the horizontal FOV doesn't increase, causing the image to look "zoomed in"), but it doesn't decrease as the viewport becomes taller.

This may hint at a configuration issue somewhere in the rendering code or shader.

**Comment by lawnjelly** [comment]
This makes sense. If your shadow map is fixed in width, in the video, as he changes the width (aspect ratio) the width of the box _relative_ to the viewport width decreases as the viewport width goes up, even though the apparent size stays the same. Thus there are less texels in the shadow map for the box in widescreen, and the observed pixellation increases.

If you changed the width and height at the same time, the box decreases / increases in size and thus the observed pixellation stays the same. The apparent size of objects is based on the height, rather than width, such that changing the aspect ratio just widens the view.

**Comment by Calinou** [comment]
Reopening as this isn't fixed in 3.2.x yet.

**Comment by Calinou** [comment]
> Hello! Is there any solution for this problem?

Yes, merge https://github.com/godotengine/godot/pull/43207 against 3.5.2 locally then build custom export templates using it :slightly_smiling_face:
